src/App.tsx

```diff
@@ -1,26 +1,225 @@
+import { FilterState } from './types/FilterState';
 
-const USER_ID = 0;
+const getFilteredTodo = (todos: Todo[], query: FilterState): Todo[] => {
+  if (query === 'All') {
+    return todos;
+  }
+
+  return todos.filter(todo => todo.completed === (query === 'Completed'));
+};
+
+function wait(delay: number): Promise<void> {
+  return new Promise(resolve => {
+    setTimeout(resolve, delay);
+  });
+}
 
 export const App: React.FC = () => {
+  const [loadingTodoId, setLoadingTodoId] = useState<Todo['id'] | null>(null);
+  const [loadingCompleted, setLoadingCompleted] = useState(false);
+  const [errorMessage, setErrorMessage] = useState<string>('');
+  const [filterState, setFilterState] = useState<FilterState>(FilterState.All);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filteredTodos, setFilteredTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [lastOperation, setLastOperation] = useState<ACTION>(ACTION.UNKNOWN);
+  const inputRef = React.useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage(MESSAGE.UNABLE_LOAD));
+  }, []);
+
+  useEffect(() => {
+    setFilteredTodos(getFilteredTodo(todos, filterState));
+  }, [filterState, todos]);
+
+  useEffect(() => {
+    if (errorMessage) {
+      const timeOutId = setTimeout(() => setErrorMessage(''), 3000);
+
+      return () => {
+        clearTimeout(timeOutId);
+      };
+    }
+  }, [errorMessage]);
+
+  useEffect(() => {
+    if ([ACTION.ADD, ACTION.DELETE].includes(lastOperation)) {
+      inputRef.current?.focus();
+    }
+  }, [lastOperation, filteredTodos, errorMessage]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const itemsLeft = todos.filter(todo => !todo.completed).length;
+
+  const onAdd = (todo: Partial<Todo>) => {
+    setErrorMessage('');
+    if (todo.title === undefined || todo.title.trim() === '') {
+      return Promise.resolve().then(() => {
+        setErrorMessage(MESSAGE.TITLE_EMPTY);
+        new Error(MESSAGE.TITLE_EMPTY);
+      });
+    }
+
+    const newTodo: Todo = {
+      id: 0,
+      completed: false,
+      userId: USER_ID,
+      ...todo,
+      title: todo.title.trim(),
+    };
+
+    setTempTodo({ ...newTodo });
+
+    if (inputRef.current) {
+      inputRef.current.disabled = true;
+    }
+
+    return postTodo(newTodo)
+      .then(serverTodo => {
+        setTodos([...todos, serverTodo]);
+      })
+      .catch(error => {
+        setErrorMessage(MESSAGE.UNABLE_ADD);
+        throw Error(error);
+      })
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+        }
+
+        setLoadingTodoId(null);
+        setTempTodo(null);
+        setLastOperation(ACTION.ADD);
+      });
+  };
+
+  const onChange = (todo: Todo, fieldsToUpdate: Partial<Todo>) => {
+    setLoadingTodoId(todo.id);
+
+    return wait(1)
```

Comment:
Remove wait here, already you have wait in request function 
![Uploading image.png…]()